File: DSA/AssignmentQuestions_8-15.py
class FirstEvenNumQ8:
    def brute_force(nums: list) -> list:
        for num in nums:
            if num % 2 == 0:
                return num
        return -1

q8Obj = FirstEvenNumQ8
print(q8Obj.brute_force([1,2,6,3,5]))
print(q8Obj.brute_force([1,7,11,3,5]))
print(q8Obj.brute_force([1,7,4,8,5]))

class SumOfElementsQ7:
    def brute_force(nums: list) -> list:
        sum = 0
        for i in nums:
            sum += i
        return sum

q7Obj = SumOfElementsQ7
# print(q7Obj.brute_force([1,2,6,3,5]))

class OddElementsQ6:
    def brute_force(nums: list) -> list:
        result = []
        for i in nums:
            if i % 2 != 0:
                result.append(i)
        return result

q6Obj = OddElementsQ6
# print(q6Obj.brute_force([1,5,6,4,3,2,8]))


class SeriesQ5:
    def brute_force(n: int) -> None:
        for _ in range(n): # rows
            for j in range(1, n+1):
                print(j, end='')
            n -= 1
            print(end='\n')
            
q5Obj = SeriesQ5
# q5Obj.brute_force(5)

class SeriesQ4:
    def brute_force(n: int) -> None:
        for i in range(1, n+1):
            for j in range(1, i+1):
                print(j, end='')
            print(end='\n')

q4Obj = SeriesQ4
# q4Obj.brute_force(5)

class MedianElementQ3:
    def brute_force(elements: list) -> int:
        length = len(elements)
        for i in range(length):
            for j in (i, length - 1):
                if elements[i] <= elements[j]:
                    continue
                else:
                    temp = elements[j]
                    elements[j] = elements[i]
                    elements[i] = temp
        print(elements)
        if length % 2 == 0:
            return elements[length//2 - 1], elements[length//2]
        else:
            return elements[length//2]
        

q3Obj = MedianElementQ3
# print(q3Obj.brute_force([10, 30, 20]))
# print(q3Obj.brute_force([10, 30, 40, 20]))


class MinOfThreeQ2:
    def brute_force(elements: list) -> int:
        temp, i = 0, 1
        while i < len(elements):
            if elements[temp] < elements[i]:
                i += 1
                continue
            else:
                temp = i
            i += 1
        
        return elements[temp]

    def builtin(elements: list) -> int:
        return min(elements)

q2Obj = MinOfThreeQ2
# print(q2Obj.builtin([3, 2, 1]))
# print(q2Obj.brute_force([1, 4, 5]))
# print(q2Obj.brute_force([4, 2, 5]))
# print(q2Obj.brute_force([4, 5, 3]))
    

class MaxOfThreeQ1:
    def brute_force(elements: list) -> int:
        temp, i = 0, 1

        while i < len(elements):
            if (elements[temp]) > elements[i]:
                i += 1
                continue
            else:
                temp = i
            i += 1
            
        return elements[temp]

    def builtin(elements: list) -> int:
        return max(elements)

q1Obj = MaxOfThreeQ1
# print(q1Obj.builtin([1, 8, 10]))
# print(q1Obj.brute_force([1, 2, 10]))
# print(q1Obj.brute_force([20, 1, 2]))
# print(q1Obj.brute_force([1, 30, 2]))

File: DSA/src/AssignmentQuestions_8-15.java
package src;

import java.util.ArrayList;
import java.util.Arrays;

class FirstEvenNumberQ8 {
    public static int bruteForce(int[] nums) {
        for (int num : nums) {
            if (num % 2 == 0)
                return num;
        }
        return -1;
    }
}

class SumOfElementsQ7 {
    public static int bruteForce(int[] nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        return sum;
    }

    public static int builtins(int[] nums) {
        return Arrays.stream(nums).sum();
    }
}

class OddElementsQ6 {
    public static ArrayList<Integer> bruteForce(int[] nums) {
        int lengthOfArray = nums.length;

        // For dynamic re-size use ArrayList :)
        ArrayList<Integer> result = new ArrayList<Integer>();

        for (int i = 0; i < lengthOfArray; i++) {
            int currentElement = nums[i];
            if (currentElement % 2 != 0) {
                result.add(currentElement);
            }
        }
        return result;
    }
}

class SeriesQ5 {
    public static void bruteForce(int n) {
        for (int i = n; i > 0; i--) {
            for (int j = 0; j < i; j++) {
                System.out.print(j + 1);
            }
            System.out.println();
        }
    }
}

class SeriesQ4 {
    public static void bruteForce(int n) {
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                System.out.print(j + 1);
            }
            System.out.println();
        }
    }
}

class MedianElementQ3 {
    public int[] bruteForce(int[] elements) {
        int lengthOfElements = elements.length;
        int temp;

        for (int i = 0; i < lengthOfElements; i++) {
            for (int j = i; j < lengthOfElements; j++) {
                if (elements[i] > elements[j]) {
                    temp = elements[i];
                    elements[i] = elements[j];
                    elements[j] = temp;
                }
            }
        }

        if (lengthOfElements % 2 == 0) {
            return new int[] { elements[lengthOfElements / 2 - 1], elements[lengthOfElements / 2] };
        }
        return new int[] { elements[lengthOfElements / 2] };
    }

    public int usingMinAndMax(int[] elements) {
        int max = new MaxOfThreeQ1().bruteForce(elements);
        int min = new MinOfThreeQ2().bruteForce(elements);
        for (int i = 0; i < elements.length; i++) {
            if (elements[i] != min && elements[i] != max) {
                return elements[i];
            }
        }
        return elements[1];
    }
}

class MinOfThreeQ2 {
    public int bruteForce(int[] elements) {
        int i = 1, temp = 0;
        while (i < elements.length) {
            if (elements[temp] > elements[i]) {
                temp = i;
            }
            i++;
        }
        return elements[temp];
    }
}

class MaxOfThreeQ1 {
    public int bruteForce(int[] elements) {
        int i = 1, temp = 0;

        while (i < elements.length) {
            if (elements[temp] < elements[i]) {
                temp = i;
            }
            i++;
        }
        return elements[temp];
    }
}

class AssignmentQuestions1 {
    public static void main(String[] args) {
        System.out.println(FirstEvenNumberQ8.bruteForce(new int[] { 1, 2, 6, 3, 5 }));
        System.out.println(FirstEvenNumberQ8.bruteForce(new int[] { 1, 7, 9, 3, 5 }));
        System.out.println(FirstEvenNumberQ8.bruteForce(new int[] { 1, 7, 4, 8, 5 }));

        // System.out.println(src.SumOfElementsQ7.builtins(new int[] { 1, 2, 6, 3, 5 }));
        // System.out.println(src.SumOfElementsQ7.bruteForce(new int[] { 1, 2, 6, 3, 5 }));

        // System.out.println(
        // src.OddElementsQ6.bruteForce(new int[] { 1, 5, 6, 4, 3, 2, 8})
        // );

        // src.SeriesQ5.bruteForce(5);

        // src.SeriesQ4.bruteForce(5);

        /*
         * int q3Median = new src.MedianElementQ3().usingMinAndMax(new int[] { 2, 5, 19 });
         * int[] q3Median = new src.MedianElementQ3().bruteForce(new int[] { 10, 30, 20, 40
         * });
         * for (int i : q3Median) {
         * System.out.println(i);
         * }
         */

        // src.MinOfThreeQ2 q2Obj = new src.MinOfThreeQ2();
        // System.out.println(q2Obj.bruteForce(new int[] { 1, 4, 5 }));
        // System.out.println(q2Obj.bruteForce(new int[] { 4, 2, 5 }));
        // System.out.println(q2Obj.bruteForce(new int[] { 4, 5, 3 }));

        // src.MaxOfThreeQ1 q1Obj = new src.MaxOfThreeQ1();
        // System.out.println(q1Obj.bruteForce(new int[] { 1, 2, 10 }));
        // System.out.println(q1Obj.bruteForce(new int[] { 20, 1, 2 }));
        // System.out.println(q1Obj.bruteForce(new int[] { 1, 30, 2 }));
        // System.out.println(q1Obj.bruteForce(new int[] { 1, 24, 124, 315246 }));
    }
}
